feat: new pkg with pkp auth helpers

sarahzdev opened the pull request:
# lit-auth-client

The new `lit-auth-client` packages provides a `LitAuthClient` class and `BaseProvider` classes to streamline authentication for PKPs.

## Features
- Handle authentication with Google and Discord via [Lit's login server](https://github.com/LIT-Protocol/lit-login-server)
- Handle authentication with Eth wallet by generating an authSig
- Handle registration and authentication via WebAuthn
- Mint and fetch PKPs using [Lit's relay server](https://github.com/LIT-Protocol/relay-server)
- Wraps `getSessionSigs` to create session sigs for a given PKP and auth method

## Changes
- Created a new browser-only package `lit-auth-client`
- In `lit-auth-client`:
  - Added new classes and methods to support auth and relay server interactions
  - Added basic tests for utility methods
  - Merged in changes from #86 
- Updated `types` and `constants` packages to support the new types
- Updated `getSessionKey` method in `lit-node-client-nodejs` (merged in #79)

## Testing

The updated packages are currently published with the `dev` tag (e.g., `yarn add @lit-protocol/lit-auth-client@dev`).

Tested all functionality in this [Next.js TypeScript app](https://github.com/LIT-Protocol/pkp-social-auth-example/pull/1). See the deployed preview [here](https://pkp-social-auth-example-4jux05hev-lit-protocol.vercel.app/).

Also tested the same functionality as above in local React, React TS, and Next.js apps.

## Next Steps

- Handle social login flow in pop-up
- Add helper functions to mint and fetch PKPs via contract calls (wrapping methods from the `contracts-sdk`)
- Error class / type

Ansonhkg commented:
It looks good to me! I've added the demo app the repo!

hwrdtm commented on apps/demo-pkp-social-auth-next-ts/package.json:
```diff
@@ -0,0 +1,31 @@
+{
```
let's try to unify to use `yarn.lock` throughout the project, no need for both `package-lock.json` and `yarn.lock`.

hwrdtm commented on apps/demo-pkp-social-auth-next-ts/src/pages/_app.tsx:
```diff
@@ -0,0 +1,45 @@
+const { chains, provider, webSocketProvider } = configureChains(
+  [mainnet],
+  [
+    alchemyProvider({ apiKey: 'onvoLvV97DDoLkAmdi0Cj7sxvfglKqDh' }),
```
should be left as a a secret / env variable

hwrdtm commented on apps/demo-pkp-social-auth-next-ts/src/pages/index.tsx:
```diff
@@ -0,0 +1,496 @@
+      }
+
+      // Get session signatures
+      const sessionSigs = await litNodeClient.getSessionSigs({
```
it might be simpler for users to not have to worry about multiple SDKs and only interact with `litAuthClient` - perhaps a light wrapper in `litAuthClient` to wrap around the `litNodeClient.getSessionSigs` call? wdyt?

hwrdtm commented on apps/demo-pkp-social-auth-next-ts/src/pages/index.tsx:
```diff
@@ -0,0 +1,496 @@
+
+    let authNeededCallback;
+    try {
+      if (authMethod.authMethodType === AuthMethodType.EthWallet) {
```
it would be great, if possible, to not have the dapp care / implement this if / else and we just implement it for them in a single `getAuthNeededCallback` call (and we do the if/else inside our SDK). Hell, it'd be cool to have the dapp not make this call at all. Might mean the SDK maintains more state. wdyt?

hwrdtm reviewed (commented):
I'd like to hear everyone's (@debbly @Ansonhkg @glitch003 @sarahzdev ) thoughts on designing an API that does **not** use the word "PKP" in favor of something abstract such as `Account`, in addition to remove / avoid exposing any notion of PKPs to the end-user / dapp / developers. The rationale is simplicity and staying intuitive for the end-user devX.

sarahzdev commented:
@Ansonhkg @hwrdtm mind giving this a re-review? since last review, i added the following:
- `LitAuthClient` now has a `litNodeClient` property
- Added `getSessionSigsWithAuth` method to `LitAuthClient` as a a convenient method to get session sigs for a given PKP and auth method
- `expiration` is optional on `SignSessionKeyProps` (`signSessionKey` falls back to a default val)
- `getSessionKey` changes from PR #79  (will dedicate PR #79 to enforcing types)

Ansonhkg commented:
Closing this because it has been merged into commit a2d58ac
